### PR TITLE
[SPARK-40105][SQL] Improve repartition in ReplaceCTERefWithRepartition

### PR DESCRIPTION
**Pull request by ulysses-you**
### What changes were proposed in this pull request?

- skip adding a repartition if the top level node of CTE is rebalance
- use RepartitionByExpression instead of Repartition so that AQE can coalesce the shuffle partition

### Why are the changes needed?

If cte can not inlined, the ReplaceCTERefWithRepartition will add repartition to force a shuffle so that the reference can reuse shuffle exchange.
The added repartition should be optimized by AQE for better performance.

If the user has specified a rebalance, the ReplaceCTERefWithRepartition should skip add repartition.

### Does this PR introduce _any_ user-facing change?

no, only improve performance

### How was this patch tested?

add test

### CODE REVIEW

**Comment by ulysses-you** [comment]
cc @cloud-fan @maryannxue

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/ReplaceCTERefWithRepartition.scala [hunk 1]
```diff
@@ -41,19 +41,25 @@ object ReplaceCTERefWithRepartition extends Rule[LogicalPlan] {
...
         val inlined = replaceWithRepartition(cteDef.child, cteMap)
         val withRepartition =
-          if (inlined.isInstanceOf[RepartitionOperation] || cteDef.underSubquery) {
+          if (canSkipExtraRepartition(inlined) || cteDef.underSubquery) {
             // If the CTE definition plan itself is a repartition operation or if it hosts a merged
             // scalar subquery, we do not need to add an extra repartition shuffle.
             inlined
           } else {
-            Repartition(conf.numShufflePartitions, shuffle = true, inlined)
+            RepartitionByExpression(Seq.empty, inlined, None)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/ReplaceCTERefWithRepartition.scala:62]: what does this no-key shuffle do eventually?

Comment (ulysses-you) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/ReplaceCTERefWithRepartition.scala:62]: it uses RoundRobinPartitioning to do exchange which is same wih Repartition

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/ReplaceCTERefWithRepartition.scala:62]: and the only difference is AQE can optimize it?

Comment (ulysses-you) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/ReplaceCTERefWithRepartition.scala:62]: yes, 

The eventually shuffle of Repartition is
`ShuffleExchangeExec(RoundRobinPartitioning, REPARTITION_BY_NUM)`

The eventually shuffle of RepartitionByExpression(Seq.empty) with none partition number is:
`ShuffleExchangeExec(RoundRobinPartitioning, REBALANCE_PARTITIONS_BY_NONE)`

File: sql/core/src/test/scala/org/apache/spark/sql/CTEInlineSuite.scala [hunk 2]
```diff
@@ -485,4 +490,23 @@ abstract class CTEInlineSuiteBase
...
+        s"""with
+           |v as (
+           |  select /*+ rebalance(c1) */ c1, c2, rand() from t
+           |)
+           |select * from v except select * from v
+         """.stripMargin)
+      checkAnswer(df, Nil)
+
+      assert(!df.queryExecution.optimizedPlan.exists(_.isInstanceOf[RepartitionOperation]))
+      assert(df.queryExecution.optimizedPlan.exists(_.isInstanceOf[RebalancePartitions]))
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/CTEInlineSuite.scala:509]: seems we are checking optimized plan here, how is it related to AQE and why we put this test only in `CTEInlineSuiteAEOn`?

Comment (ulysses-you) [reply sql/core/src/test/scala/org/apache/spark/sql/CTEInlineSuite.scala:509]: rebalance hint can only be resolved with AQE on, so this test does not work if AQE is disabled.

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by ulysses-you** [comment]
thank you @cloud-fan

**Pull request closed without merging**